File: agents/agent-dev/agent-patterns/agent-type-patterns/react/prebuilt-agent-ts-poc/agent.ts
// agent.ts

import { DuckDuckGoSearch } from "@langchain/community/tools/duckduckgo_search";
import { ChatOpenAI } from "@langchain/openai";
import { MemorySaver } from "@langchain/langgraph";
import { HumanMessage } from "@langchain/core/messages";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import * as dotenv from 'dotenv';
dotenv.config();

console.log("Starting agent...");

const apiKey = process.env.OPENAI_API_KEY;
if (!apiKey) {
  throw new Error("Missing OPENAI_API_KEY in environment variables");
}

// Define the tools for the agent to use
const agentTools = [new DuckDuckGoSearch({ maxResults: 3 })];

const agentModel = new ChatOpenAI(
  { model:"gpt-4",
    temperature: 0,
    apiKey: apiKey
   }
);

// Initialize memory to persist state between graph runs
const agentCheckpointer = new MemorySaver();
const agent = createReactAgent({
  llm: agentModel,
  tools: agentTools,
  checkpointSaver: agentCheckpointer,
});

main()

async function main() {

// Now it's time to use!
const agentFinalState = await agent.invoke(
  { messages: [new HumanMessage("what is the current weather in sf")] },
  { configurable: { thread_id: "42" } },
);

console.log(
  agentFinalState.messages[agentFinalState.messages.length - 1].content,
);

const agentNextState = await agent.invoke(
  { messages: [new HumanMessage("what about ny")] },
  { configurable: { thread_id: "42" } },
);

console.log(
  agentNextState.messages[agentNextState.messages.length - 1].content,
);

}

File: ops/ci-cd/actions/action-poc/agent.py
# agent.py
from langchain_openai import ChatOpenAI
from langchain_core.tools import tool
from langgraph.prebuilt import create_react_agent

# Define a simple tool
@tool
def add(x: int, y: int) -> int:
    """Add two numbers together."""
    return x + y

tools = [add]

# Set up an LLM (requires OPENAI_API_KEY in env)
model = ChatOpenAI(model="gpt-4o-mini", temperature=0)

# Create ReAct agent with tools
agent = create_react_agent(model, tools)

if __name__ == "__main__":
    # Run one interaction
    result = agent.invoke(
        {"messages": [{"role": "user", "content": "What is 2 + 3?"}]}
    )
    print(result["messages"][-1].content)
